### PR TITLE
feat: lazy materialize according to virtual column `_row_id`.

### LINKED ISSUE #11126: bug: TopN query with `_row_id` will panic if the storage is native format in debug build.
**Summary**

```sql
select b, _row_id from t order by a limit 10; -- storage of `t` is native format.
```

### LINKED ISSUE #11138: bug: tpcds load_data.sh typo
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

https://github.com/datafuselabs/databend/tree/main/benchmark/tpcds

### What's Wrong?

the env.sh in the file should be shell_env.sh

### How to Reproduce?

open the file and search env.sh

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by RinChanNOWWW**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- Add a new transform `TransformNativeRowsFetcher`. It fetches and deserializes specific columns.
- Analyze if a query can be optimized by lazy materialization in binder. Currently, only simple Top-N query on Fuse engine can be optimized. (single table, no aggregation, no cte).

```sql
select ... from t ... where ... order by ... limit ... ;
```

- Add a new setting value `lazy_top_threshold` to control the behavior: Enable lazy materialization and set the limit threshold of Top-N queries. Set the value to 0 to disable this setting.
- Fix #11126
- Fix typo #11138 

Closes #11126
Closes #11138

### CODE REVIEW

File: src/query/catalog/src/table.rs [hunk 1]
```diff
@@ -110,8 +110,13 @@ pub trait Table: Sync + Send {
         false
     }
 
-    /// Whether the table engine supports virtual columns optimization.
-    fn support_virtual_columns(&self) -> bool {
+    /// Whether the table engine supports virtual columns optimization for Variant types.
+    fn support_variant_virtual_columns(&self) -> bool {
```
Comment (b41sh) [inline src/query/catalog/src/table.rs:114]: Virtual columns can be used not only for variant type, but also for internal columns of Map type or other nested types.

Comment (RinChanNOWWW) [reply src/query/catalog/src/table.rs:114]: I want to distinguish row_id and other virtual columns.

Comment (b41sh) [reply src/query/catalog/src/table.rs:114]: To distinguish it from a `VirtualColumn`, the `row_id` can be treated as an `InternalColumn`, for example in the definition of `ColumnEntry`. https://github.com/datafuselabs/databend/blob/main/src/query/sql/src/planner/metadata.rs#L413

File: src/query/catalog/src/table_context.rs [hunk 2]
```diff
@@ -101,6 +101,9 @@ pub trait TableContext: Send + Sync {
     fn get_cacheable(&self) -> bool;
     fn set_cacheable(&self, cacheable: bool);
 
+    fn set_snapshot(&self, location: String, ver: u64);
+    fn get_snapshot(&self) -> Option<(String, u64)>;
```
Comment (lichuang) [inline src/query/catalog/src/table_context.rs:105]: I think type `Location` is better to represent `location: String, ver: u64` tuple

File: src/query/sql/src/planner/binder/select.rs [hunk 3]
```diff
@@ -667,6 +693,96 @@ impl Binder {
...
+        select_list: &SelectList,
+        where_scalar: &Option<ScalarExpr>,
+        order_by: &[OrderItem],
+        limit: usize,
+    ) -> Result<()> {
+        // Only simple single table Top-N query is supported.
+        // e.g.
+        // SELECT ... FROM t WHERE ... ORDER BY ... LIMIT ...
+        if stmt.group_by.is_some()
+            || stmt.having.is_some()
```
Comment (sundy-li) [inline src/query/sql/src/planner/binder/select.rs:712]: no `unnest`

File: src/query/service/src/pipelines/pipeline_builder.rs [hunk 4]
```diff
@@ -992,6 +1006,24 @@ impl PipelineBuilder {
         })
     }
 
+    fn build_row_fetch(&mut self, row_fetch: &RowFetch) -> Result<()> {
+        debug_assert!({
```
Comment (zhang2014) [inline src/query/service/src/pipelines/pipeline_builder.rs:1010]:
```suggestion
        debug_assert!(matches!(&*row_fetch.input, PhysicalPlan::Limit(limit) if matches!(*limit.input, PhysicalPlan::Sort(_)));
```

**Review by zhang2014 — APPROVED** [review]

File: src/query/storages/fuse/src/operations/read/fuse_row_fetcher.rs [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            ver,
+            self.fetcher.reader().operator.clone(),
+            self.table_schema.clone(),
+            &self.projection,
+        )
+        .await?;
+        Ok(())
+    }
+
+    // #[async_backtrace::framed]
```
Comment (b41sh) [inline src/query/storages/fuse/src/operations/read/fuse_row_fetcher.rs:220]: unused comment?

Comment (RinChanNOWWW) [reply src/query/storages/fuse/src/operations/read/fuse_row_fetcher.rs]: uncommented.

**Review by sundy-li — APPROVED** [review]

File: src/query/sql/src/executor/format.rs [hunk 6]
```diff
@@ -602,6 +604,40 @@ fn limit_to_format_tree(
     Ok(FormatTreeNode::with_children("Limit".to_string(), children))
 }
 
+fn row_fetch_to_format_tree(
```
Comment (b41sh) [inline src/query/sql/src/executor/format.rs:607]: How about adding an `explain` test case?

**Comment by RinChanNOWWW** [comment]
@mergify refresh

**Comment by RinChanNOWWW** [comment]
Thank @zhang2014 very very much for helping to solve the stack overflow issue!!!

**Pull request merged**